File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    bool canship(vector<int>& weights, int cap, int days) {
        int load = 0;
        int useddays = 1;

        for (int w : weights) {
            if (load + w > cap) {
                useddays++;
                load = 0;
            }
            load += w;
        }
        return useddays <= days;
    }

    int shipWithinDays(vector<int>& weights, int days) {
        int low = *max_element(weights.begin(), weights.end());
        int high = 0;
        for (int w : weights) high += w;

        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (canship(weights, mid, days))
                high = mid - 1;
            else
                low = mid + 1;
        }
        return low;
    }
};

File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.java
class Solution {
    public int canship(int[] w,int mid){
        int days=0;
        int sum=0;
        for(int i:w){
            if(sum+i>mid){
                sum=0;
                days++;
            }
            sum+=i;
        }
        days++;
        return days;
    }
    public int shipWithinDays(int[] weights, int days) {
        int n=weights.length;
        int low=1;
        int high=0;
        for(int i:weights){
            low=Math.max(low,i);
            high+=i;
        }

        int ans=0;
        while(low<=high){
            int mid=low+(high-low)/2;
            if(canship(weights,mid)<=days){
                ans=mid;
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }
        return ans;

    }
}